subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/RepositoryChainComponentMetaDataResolver.java

```diff
@@ -79,54 +79,66 @@ public boolean isFetchingMetadataCheap(ComponentIdentifier identifier) {
 
+        List<Throwable> errors = new ArrayList<Throwable>();
+
         List<ComponentMetaDataResolveState> resolveStates = new ArrayList<ComponentMetaDataResolveState>();
         for (ModuleComponentRepository repository : repositories) {
             resolveStates.add(new ComponentMetaDataResolveState(identifier, componentOverrideMetadata, repository, versionedComponentChooser));
         }
 
-        try {
-            final RepositoryChainModuleResolution latestResolved = findBestMatch(resolveStates);
-            if (latestResolved == null) {
-                for (ComponentMetaDataResolveState resolveState : resolveStates) {
-                    resolveState.applyTo(result);
-                }
-                result.notFound(identifier);
-            } else {
-                LOGGER.debug("Using {} from {}", latestResolved.module.getId(), latestResolved.repository);
-                result.resolved(metaDataFactory.transform(latestResolved));
+        final RepositoryChainModuleResolution latestResolved = findBestMatch(resolveStates, errors);
+        if (latestResolved != null) {
+            LOGGER.debug("Using {} from {}", latestResolved.module.getId(), latestResolved.repository);
+            for (Throwable error : errors) {
+                LOGGER.debug("Discarding resolve failure.", error);
             }
-        } catch (Throwable error) {
-            result.failed(new ModuleVersionResolveException(identifier, error));
+
+            result.resolved(metaDataFactory.transform(latestResolved));
+            return;
+        }
+        if (!errors.isEmpty()) {
+            result.failed(new ModuleVersionResolveException(identifier, errors));
+        } else {
+            for (ComponentMetaDataResolveState resolveState : resolveStates) {
+                resolveState.applyTo(result);
+            }
+            result.notFound(identifier);
         }
     }
 
-    private RepositoryChainModuleResolution findBestMatch(List<ComponentMetaDataResolveState> resolveStates) throws Throwable {
+    private RepositoryChainModuleResolution findBestMatch(List<ComponentMetaDataResolveState> resolveStates, Collection<Throwable> failures) {
         LinkedList<ComponentMetaDataResolveState> queue = new LinkedList<ComponentMetaDataResolveState>();
         queue.addAll(resolveStates);
 
         LinkedList<ComponentMetaDataResolveState> missing = new LinkedList<ComponentMetaDataResolveState>();
 
         // A first pass to do local resolves only
-        RepositoryChainModuleResolution best = findBestMatch(queue, missing);
+        RepositoryChainModuleResolution best = findBestMatch(queue, failures, missing);
         if (best != null) {
             return best;
         }
 
-        // Nothing found locally - try a remote search for all resolve states that were not yet searched remotely
+        // Nothing found - do a second pass
```

Comment:
Why change this informative comment? Is it wrong?
I think the algorithm here is complex, and requires more documentation, not less.

Reply:
I think Ben just reverted Jendrik's commit, accidentally also reverting this helpful line.